llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp

```diff
@@ -0,0 +1,432 @@
+  FunctionType *FTy = F.getFunctionType();
+  Type *RetTy = FTy->getReturnType();
+  SmallVector<Type *, 8> ArgTys;
+  AttributeList AttrVec;
+  const AttributeList &PAL = F.getAttributes();
+  // Argument, result of load
+  DenseMap<Argument *, Value *> ToModify;
+  auto &Context = F.getContext();
+
+  for (unsigned int i = 0; i != F.arg_size(); i++) {
+    auto Arg = F.getArg(i);
+    Type *NewTy = argIsSimdPtr(Arg);
+    if (NewTy) {
+      // Copy over byval type for simd* type
+      ArgTys.push_back(NewTy);
+    } else {
+      // Transfer all non-simd ptr arguments
+      ArgTys.push_back(Arg->getType());
+      AttributeSet Attrs = PAL.getParamAttributes(i);
+      if (Attrs.hasAttributes()) {
+        AttrBuilder B(Attrs);
+        AttrVec = AttrVec.addParamAttributes(Context, i, B);
+      }
+    }
+  }
+
+  FunctionType *NFTy = FunctionType::get(RetTy, ArgTys, false);
+
+  // Add any function attributes
+  AttributeSet FnAttrs = PAL.getFnAttributes();
+  if (FnAttrs.hasAttributes()) {
+    AttrBuilder B(FnAttrs);
+    AttrVec = AttrVec.addAttributes(Context, AttributeList::FunctionIndex, B);
+  }
+
+  auto RetAttrs = PAL.getRetAttributes();
+  if (RetAttrs.hasAttributes()) {
+    AttrBuilder B(RetAttrs);
+    AttrVec = AttrVec.addAttributes(Context, AttributeList::ReturnIndex, B);
+  }
+
+  // Create new function body and insert into the module
+  Function *NF = Function::Create(NFTy, F.getLinkage(), F.getName());
+  NF->copyAttributesFrom(&F);
+  NF->setCallingConv(F.getCallingConv());
+
+  F.getParent()->getFunctionList().insert(F.getIterator(), NF);
+  NF->takeName(&F);
+  NF->setSubprogram(F.getSubprogram());
+
+  // Now to splice the body of the old function into the new function
+  NF->getBasicBlockList().splice(NF->begin(), F.getBasicBlockList());
+
+  for (unsigned int I = 0; I != F.arg_size(); I++) {
+    auto Arg = F.getArg(I);
+    Type *newTy = argIsSimdPtr(Arg);
+    if (newTy) {
+      // Insert bitcast
+      // bitcast vector* -> simd*
+      auto BitCast = new BitCastInst(NF->getArg(I), Arg->getType());
+      NF->begin()->getInstList().push_front(BitCast);
+      ToModify.insert(std::make_pair(Arg, nullptr));
+      Arg->replaceAllUsesWith(BitCast);
+    }
+  }
+
+  // Loop over the argument list, transferring uses of the old arguments to the
+  // new arguments, also tranferring over the names as well
+  Function::arg_iterator I2 = NF->arg_begin();
+  unsigned int ArgNo = 0;
+  for (Function::arg_iterator I = F.arg_begin(), E = F.arg_end(); I != E;
+       ++I, ++I2, ArgNo++) {
+    auto ArgIt = ToModify.find(I);
+    if (ArgIt == ToModify.end()) {
+      // Transfer old arguments as is
+      I->replaceAllUsesWith(I2);
+      I2->takeName(I);
+    }
+  }
+
+  // Fix call sites
+  SmallVector<std::pair<Instruction *, Instruction *>, 10> OldNewInst;
+  for (auto &use : F.uses()) {
+    // Use must be a call site
+    SmallVector<Value *, 10> Params;
+    auto User = use.getUser();
+    if (isa<CallInst>(User)) {
```

Comment:
Do you need to handle InvokeInst also?

Reply:
from what i read, InvokeInst is used for exception handling. for device code, there is no exception handling support.

Reply:
Yes, that's true.